### ISSUE TITLE
Subscribe and receive last message as first?

### ISSUE THREAD

**Issue by memelet** [allegro/hermes#761]
Is it possible to cache the last messsage sent on a topic such that new subscribers get that message immediately on subscribe? This is a common pattern that has been implemented in most brokers to support topics that represent state.

**Comment by adamdubiel** [comment]
When you subscribe, you get all new messages that were published after subscribing. As far as i understand you would like to get the last message published right before creating subscription?

I think it doesn't matter much when you have events published often (like 1 event per second or more). I can only imagine this being useful when publication rate on topic is very low (like one event for every few minutes). Do i understand correctly?

**Comment by memelet** [comment]
Yes exactly. It's for state like values that change slowly. But this feature was implemented in activemq ten years ago to support an options trading system I worked on. In that case the data changed fast but retaining the last value in the broker drastically reduced latency where the last value was needed sooner than milliseconds from subscribers.

**Comment by adamdubiel** [comment]
Under the hood we rely on Kafka subscription model so we would need to check if it is possible to simply pass smaller offset number to Kafka to receive messages published earlier.

However creating some kind of "in memory cache" would be really expensive and i see a lot of hard to solve corner cases. We rely on Kafka offset reading & committing, and it would be hard to manage retries between Consumer restarts.

Also in Kafka each subscription keeps own position (offset) in topic. So there is no "latest message". To cache this kind of thing, we would have to create an artificial subscription reading and discarding events just to keep the latest one in memory on all Consumers hosts. In our use cases this would be a waste of resources.

Summing up: it probably can be achieved, but at a huge cost, and it is currently out of scope for us. If you are still interested in this feature and can code it yourself, we can help with code and release it.

**Comment by memelet** [comment]
Fair enough. I understand that kafka is not quite a fit for this model, but I was not to sure how to what extent hermes layers on top and whether not this would be feasible. 

Indeed with systems like activemq and rabbitmq and some of the mqtt brokers that support this managed a lot more state than kafka, and for sure need to store the "last values" in memory or in their persistent structures.
